app/controllers/ApiController.scala

```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
     */
   def showStatusZ = Action(Ok(this.status.json))
 
+  def syncSso() = Action { implicit request =>
+    this.forms.SyncSso.bindFromRequest.fold(
+      hasErrors => BadRequest(Json.obj("errors" -> hasErrors.errorsAsJson)),
+      success = formData => {
+        val sso = formData._1
+        val sig = formData._2
+        val apiKey = formData._3
+
+
+        val confApiKey = this.config.security.get[String]("sso.apikey")
+        val confSecret = this.config.security.get[String]("sso.secret")
+
+        if (apiKey != confApiKey) {
+          BadRequest("API Key not valid")
+        } else if (!CryptoUtils.hmac_sha256(confSecret, sso.getBytes("UTF-8")).equals(sig)) {
+          BadRequest("Signature not matched")
+        } else {
+          val query = Uri.Query(Base64.getMimeDecoder.decode(sso))
+
+          val external_id = query.get("external_id")
+          val email = query.get("email")
+          val username = query.get("username")
+          val name = query.get("name")
+          val avatar_url = query.get("avatar_url")
+
+          val optUser = this.users.get(external_id.getOrElse("-1").toInt)
+
+          if (optUser.isEmpty) {
```

Comment:
This statement does nothing in this position. Also not a fan of testing the content in this way. I'd say the best here would be a pattern match on optUser.